rspace/src/main/scala/coop/rchain/rspace/history/package.scala

```diff
@@ -230,22 +373,54 @@ package object history {
           case Vector() => throw new DeleteException("PointerBlock has no children")
           // If there are is only one child, then we know that it is the thing we are trying to
           // propagate upwards, and we can go ahead and do that.
-          case Vector(_) => propagateLeafUpward(hash, tail)
+          case Vector(_) =>
+            propagateTrieUpward(store, txn, ptr, ByteVector(byte), tail)
           // Otherwise, if there are > 2 children, we can update the parent node's Vector
           // at the given index to point to the leaf.
-          case _ => (Node(pointerBlock.updated(List((byte, LeafPointer(hash))))), tail)
+          case _ =>
+            val setPtr = if (incomingAffix.length > 0) {
+              setupSkipNode(store, txn, ptr, incomingAffix)
+            } else {
+              ptr
+            }
+            (Node(pointerBlock.updated(List((byte, setPtr)))), tail)
         }
     }
 
+  private[this] def setupSkipNode[V, K, T](
+      store: ITrieStore[T, K, V],
+      txn: T,
+      ptr: NonEmptyPointer,
+      incomingAffix: ByteVector)(implicit codecK: Codec[K], codecV: Codec[V]) = {
+    val skip     = Skip(incomingAffix, ptr)
+    val skipHash = Trie.hash(skip)(codecK, codecV)
+    store.put(txn, skipHash, skip)
```

Comment:
Seems like there is nothing that requires us to `put` here, so long as we hold on to the `skip`.  We could then keep something like this pure and batch insert elsewhere.